daemon/api_validate.go

```diff
@@ -0,0 +1,260 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+type validationSetResult struct {
+	ValidationSet string `json:"validation-set,omitempty"`
+	Mode          string `json:"mode"`
+	Seq           int    `json:"seq,omitempty"`
+	Valid         bool   `json:"valid"`
+	// TODO: attributes for Notes column
+}
+
+func modeString(mode assertstate.ValidationSetMode) (string, error) {
+	switch mode {
+	case assertstate.Monitor:
+		return "monitor", nil
+	case assertstate.Enforce:
+		return "enforce", nil
+	}
+	return "", fmt.Errorf("internal error: unhandled mode %d", mode)
+}
+
+func validationSetKeyString(accountID, name string, sequence int) string {
+	key := assertstate.ValidationSetKey(accountID, name)
+	if sequence != 0 {
+		key = fmt.Sprintf("%s=%d", key, sequence)
+	}
+	return key
+}
+
+func validationSetNotFound(accountID, name string, sequence int) Response {
+	res := &errorResult{
+		Message: "validation set not found",
+		Kind:    client.ErrorKindValidationSetNotFound,
+		Value:   validationSetKeyString(accountID, name, sequence),
+	}
+	return &resp{
+		Type:   ResponseTypeError,
+		Result: res,
+		Status: 404,
+	}
+}
+
+func validationSetAccountAndName(validationSet string) (accountID, name string, err error) {
+	parts := strings.Split(validationSet, "/")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid validation-set argument")
+	}
+	accountID = parts[0]
+	name = parts[1]
+	if !asserts.IsValidAccountID(accountID) {
+		return "", "", fmt.Errorf("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return "", "", fmt.Errorf("invalid validation set name %q", name)
+	}
+	return accountID, name, nil
+}
+
+func listValidationSets(c *Command, r *http.Request, _ *auth.UserState) Response {
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	validationSets, err := assertstate.ValidationSets(st)
+	if err != nil {
+		return InternalError("accessing validation sets failed: %v", err)
+	}
+
+	names := make([]string, 0, len(validationSets))
+	for k := range validationSets {
+		names = append(names, k)
+	}
+	sort.Strings(names)
+
+	results := make([]validationSetResult, len(names))
+	for i, vs := range names {
+		tr := validationSets[vs]
+		// TODO: evaluate against installed snaps
+		var valid bool
+		modeStr, err := modeString(tr.Mode)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+		results[i] = validationSetResult{
+			ValidationSet: validationSetKeyString(tr.AccountID, tr.Name, tr.PinnedAt),
```

Comment:
I missed that we would incorporate pinned-at in the string itself, I wonder if we should make it a separate field instead

Reply:
Yes I think separate fields also for account, name and pinned-at make sense.